### ISSUE TITLE
[web] CanvasKit render special characters

### ISSUE THREAD

**Issue by YQYOnTheRoad** [flutter/flutter#53897]
I use command     "flutter run -d chrome --release"
![image](https://user-images.githubusercontent.com/19703102/78324402-35c03a80-75a7-11ea-98da-9e2f560bbb00.png)

 I use command     "flutter run -d chrome --dart-define=FLUTTER_WEB_USE_SKIA=true --release
![image](https://user-images.githubusercontent.com/19703102/78324428-47094700-75a7-11ea-998f-7a9ea7c81891.png)

flutter doctor -v 
[✓] Flutter (Channel dev, v1.16.4-pre.33, on Mac OS X 10.15.4 19E266, locale
    zh-Hans-CN)
    • Flutter version 1.16.4-pre.33 at /Users/ybc/Downloads/flutter
    • Framework revision aee9e94c21 (3 days ago), 2020-03-30 16:12:40 -0700
    • Engine revision c779894d08
    • Dart version 2.8.0 (build 2.8.0-dev.17.0 1402e8e1a4)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/ybc/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/ybc/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4, Build version 11E146
    • CocoaPods version 1.9.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.43.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.1

[✓] Connected device (3 available)
    • MP1718     • 437dac39   • android-arm64  • Android 8.1.0 (API 27)
    • Chrome     • chrome     • web-javascript • Google Chrome 80.0.3987.163
    • Web Server • web-server • web-javascript • Flutter Tools

• No issues found!

it's a new project， I have n’t changed anything, 
English letters are normal，But Chinese garbled

**Comment by VladyslavBondarenko** [comment]
<details>
 <summary>minimal code sample to reproduce</summary>

```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Text('一些簡短的文字'),
      ),
    );
  }
}
```

</details>

**Comment by YQYOnTheRoad** [comment]
> minimal code sample to reproduce
> Related to #47041
> @YQYOnTheRoad please follow up on the issue

I tried PonyCui's method in #47041 ，Successfully displayed the text
But text padding doesn't work anymore  ，Some icons disappear
So this is not a perfect solution
This problem is already on your agenda, right?
I will keep following，thanks

**Comment by VladyslavBondarenko** [comment]
@YQYOnTheRoad 
no, It's not in my agenda, when somebody start to work on the issue, it will be assigned to him

**Comment by yjbanov** [comment]
This may be a duplicate of https://github.com/flutter/flutter/issues/54330

**Comment by AngusPei** [comment]
Reproducible on the dev 1.21.0-9.0.pre

<details>
  <summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel dev, 1.21.0-9.0.pre, on macOS 11.0 20A5343i, locale
    zh-Hans-CN)
    • Flutter version 1.21.0-9.0.pre at /Users/angus/flutter
    • Framework revision 7c6f9dd239 (11 days ago), 2020-08-09 10:31:03 -0400
    • Engine revision 6d86e67f04
    • Dart version 2.10.0 (build 2.10.0-4.0.dev 0341576448)
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/angus/Library/Android/sdk
    • Platform android-R, build-tools 29.0.2
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.0.beta.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2019.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 43.0.3
    • Dart plugin version 193.6015.53

[✓] VS Code (version 1.48.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (3 available)
    • macOS (desktop)  • macos      • darwin-x64     • macOS 11.0 20A5343i
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome
      84.0.4147.125
```

**Comment by markusaksli-nc** [comment]
Reproducible on the latest master `1.22.0-10.0.pre.87`.
<details>
 <summary>flutter doctor -v</summary>

```
[√] Flutter (Channel master, 1.22.0-10.0.pre.87, on Microsoft Windows [Version 10.0.19041.450], locale en-US)
    • Flutter version 1.22.0-10.0.pre.87 at C:\Development\flutter_master
    • Framework revision 4732a214a7 (2 days ago), 2020-09-04 21:05:02 -0400
    • Engine revision ac8b9c4c52
    • Dart version 2.10.0 (build 2.10.0-86.0.dev)


[√] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at C:\Users\marku\AppData\Local\Android\sdk
    • Platform android-30, build-tools 30.0.1
    • Java binary at: C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.6.30320.27
    • Windows 10 SDK version 10.0.18362.0

[√] Android Studio (version 4.0)
    • Android Studio at C:\Users\marku\AppData\Local\JetBrains\Toolbox\apps\AndroidStudio\ch-0\193.6626763
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] Connected device (5 available)
    • SM G950F (mobile)           • ce12171c51cc001c03 • android-arm64  • Android 9 (API 28)
    • sdk gphone x86 arm (mobile) • emulator-5554      • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)           • windows            • windows-x64    • Microsoft Windows [Version 10.0.19041.450]
    • Web Server (web)            • web-server         • web-javascript • Flutter Tools
    • Chrome (web)                • chrome             • web-javascript • Google Chrome 85.0.4183.83

• No issues found!
```
</details>

**Comment by kunit1** [comment]
@hterkelsen This bug also affects emojis on CanvasKit. How would the proposed solution solve that issue?

**Comment by steffenhaak** [comment]
For everybody searching for an emoji/special chars workaround that works with the current CanvasKit implementation: you can bundle "Noto Sans Symbols" and "Noto Color Emoji" fonts with your flutter app and set those fonts as fontFamilyFallback in your app's theme or wherever applicable. Chinese characters should work the same way by using Noto Sans CJK, though I have not tried this myself. Naturally this increases download size by quite a lot, but unless @hterkelsen finds a way to load the system's standard font (probably prohibited due to sandboxing), I guess there is no way around anyways.

Hope this helps somebody. We struggled quite a while and are pretty happy to have found this as a working solution :)

**Comment by creativecreatorormaybenot** [comment]
@hterkelsen Can you elaborate on why it is not possible to load the system fonts for emojis currently and more importantly what this means for the outlook for this issue?

**Comment by creativecreatorormaybenot** [comment]
@hterkelsen Thanks for that amazing insight 👌

Can you elaborate on why it is not possible to load the system fonts for emojis / special characters currently and more importantly what this means for the outlook for this issue?

**Comment by ReniDelonzek** [comment]
I noticed that when the text is selected inside an input, the emoji becomes visible

**Comment by sunjianan9900** [comment]
> For everybody searching for an emoji/special chars workaround that works with the current CanvasKit implementation: you can bundle "Noto Sans Symbols" and "Noto Color Emoji" fonts with your flutter app and set those fonts as fontFamilyFallback in your app's theme or wherever applicable. Chinese characters should work the same way by using Noto Sans CJK, though I have not tried this myself. Naturally this increases download size by quite a lot, but unless @hterkelsen finds a way to load the system's standard font (probably prohibited due to sandboxing), I guess there is no way around anyways.
> 
> Hope this helps somebody. We struggled quite a while and are pretty happy to have found this as a working solution :)

thanks！
can it solved by using google font or something？
Loading the single character from web to reduse the size

**Comment by sunjianan9900** [comment]
I add this on web/index.html <head>

<link href="https://fonts.googleapis.com/css2?family=Zhi+Mang+Xing&display=swap" rel="stylesheet">    <style>
body{ font-family: 'ZCOOL QingKe HuangYou', sans-serif;}
</style>

and in main.dart

theme: ThemeData(
        fontFamily: "ZCOOL QingKe HuangYou",
)

it's work on FLUTTER_WEB_USE_SKIA=false , but it's not work on FLUTTER_WEB_USE_SKIA=true model

**Comment by truthmast** [comment]
> I noticed that when the text is selected inside an input, the emoji becomes visible

You're probably seeing text from the transparent HTML textarea that Flutter uses. Usually you can't highlight text in there, but there are a couple ways to end up focusing the textarea. Most commonly this ends up happening when using the browser's find [`ctrl + f`] text function).

**Comment by IchordeDionysos** [comment]
@hterkelsen wouldn't the [local fonts API](https://web.dev/local-fonts/) help here?
So instead of bundling a font as a fallback Flutter for the web could load a local font, right?

Of course, it will be a long way until this API is supported in all browsers.

**Comment by creativecreatorormaybenot** [comment]
@hterkelsen What is the reason that "CanvasKit cannot access system fonts"?

I was not able to find this problem with any other framework.

**Comment by ReniDelonzek** [comment]
Any update on this problem? It is a major problem for the use of canvasKit, but it seems that the Flutter team is not paying much attention 😕

**Comment by hterkelsen** [comment]
I am currently working on this. Our solution involves downloading Noto fonts when we detect that a character isn't supported by any of the given fonts.

**Comment by creativecreatorormaybenot** [comment]
@hterkelsen That is super awesome news! 🚀 🎊 

Was just about to implement it in a package myself :D

**Comment by ZhangLiuCheng** [comment]
@hterkelsen 
1. Loading custom fonts can solve the problem of messy code, but the font is relatively large, which affects the speed of the first load.
2. Using the Google_fonts in pub. dev can also be solved with long fonts, but goole_fonts will be briefly scrambled before they can be loaded.  If the text is not displayed when the code is scrambled, this can be fixed.

**Comment by creativecreatorormaybenot** [comment]
@ZhangLiuCheng Noto are not part of Google Fonts.

**Comment by ctrysbita** [comment]
![image](https://user-images.githubusercontent.com/8357481/103334366-f7ab6500-4aab-11eb-92d5-2e7426e9b6ce.png)
@hterkelsen I've tried to run Flutter Gallery web using the latest master channel. But it send thousands of requests for fonts.

**Comment by ReniDelonzek** [comment]
> @hterkelsen I've tried to run Flutter Gallery web using the latest master channel. But it sent thousands of requests for fonts which made my browser slow. Maybe we should cache these font units?

I noticed that too. This occurs because there is a character-by-character check to check whether the font exists or not.

The solutions I imagine are:

1. Maintain a queue of requests that are happening until they reach the end, and before starting a new one, check that the source is not already being downloaded.

2. Adopt cache strategies

**Comment by yjbanov** [comment]
Reopening as https://github.com/flutter/engine/pull/23096 was reverted

**Comment by creativecreatorormaybenot** [comment]
@yjbanov What is the status on that? Was it a minor issue that will be resolved or was there a fundamental issue?

**Comment by Mecil9** [comment]
> I use command "flutter run -d chrome --release"
> ![image](https://user-images.githubusercontent.com/19703102/78324402-35c03a80-75a7-11ea-98da-9e2f560bbb00.png)
> 
> I use command "flutter run -d chrome --dart-define=FLUTTER_WEB_USE_SKIA=true --release
> ![image](https://user-images.githubusercontent.com/19703102/78324428-47094700-75a7-11ea-998f-7a9ea7c81891.png)
> 
> flutter doctor -v
> [✓] Flutter (Channel dev, v1.16.4-pre.33, on Mac OS X 10.15.4 19E266, locale
> zh-Hans-CN)
> • Flutter version 1.16.4-pre.33 at /Users/ybc/Downloads/flutter
> • Framework revision [aee9e94](https://github.com/flutter/flutter/commit/aee9e94c21009bfc6c08f442eacde06f001c25f9) (3 days ago), 2020-03-30 16:12:40 -0700
> • Engine revision [c779894](https://github.com/flutter/flutter/commit/c779894d089554620134408b3cb6ea515489dc14)
> • Dart version 2.8.0 (build 2.8.0-dev.17.0 1402e8e1a4)
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
> • Android SDK at /Users/ybc/Library/Android/sdk
> • Android NDK location not configured (optional; useful for native profiling support)
> • Platform android-28, build-tools 28.0.3
> • ANDROID_HOME = /Users/ybc/Library/Android/sdk
> • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
> • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
> • All Android licenses accepted.
> 
> [✓] Xcode - develop for iOS and macOS (Xcode 11.4)
> • Xcode at /Applications/Xcode.app/Contents/Developer
> • Xcode 11.4, Build version 11E146
> • CocoaPods version 1.9.1
> 
> [✓] Chrome - develop for the web
> • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
> 
> [✓] Android Studio (version 3.6)
> • Android Studio at /Applications/Android Studio.app/Contents
> • Flutter plugin version 44.0.2
> • Dart plugin version 192.7761
> • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
> 
> [✓] VS Code (version 1.43.2)
> • VS Code at /Applications/Visual Studio Code.app/Contents
> • Flutter extension version 3.8.1
> 
> [✓] Connected device (3 available)
> • MP1718 • 437dac39 • android-arm64 • Android 8.1.0 (API 27)
> • Chrome • chrome • web-javascript • Google Chrome 80.0.3987.163
> • Web Server • web-server • web-javascript • Flutter Tools
> 
> • No issues found!
> 
> it's a new project， I have n’t changed anything,
> English letters are normal，But Chinese garbled

这个问题你解决了吗？VScode出现这个问题，AS可以正常显示，我这里

**Comment by pcsosinski** [comment]
@yjbanov please update to an appropriate milestone

**Comment by YowFung** [comment]
Same issue in flutter 2.x, when Chinese characters are followed by an `overflow: TextOverflow.ellipsis`.

![image](https://user-images.githubusercontent.com/26346560/117234020-caeecc00-ae56-11eb-909f-defb2b003103.png)

<details>

<summary>flutter doctor -v</summary>

```txt
[√] Flutter (Channel stable, 2.0.6, on Microsoft Windows [Version 10.0.17763.1817], locale zh-CN)
    • Flutter version 2.0.6 at D:\SDK\Flutter
    • Framework revision 1d9032c7e1 (6 days ago), 2021-04-29 17:37:58 -0700
    • Engine revision 05e680e202
    • Dart version 2.12.3
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at D:\SDK\Android
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = D:\SDK\Android
    • ANDROID_SDK_ROOT = D:\SDK\Android
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.8.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.8.30907.101
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] VS Code (version 1.55.2)
    • VS Code at C:\Users\Zhanghuifeng\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.17763.1817]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 90.0.818.51

• No issues found!
```

</details>

**Comment by mdrideout** [comment]
Hey All,

Just checking in on this. I am currently working on a Flutter Web app which must support English, Japanese, Korean, and a variety of other languages. My current approach will be to dynamically load a Google Font of the supporting language based on the selected locale (using [easy_localization](https://pub.dev/packages/easy_localization) package for localization), and re-render the app with the updated font. I will post my solution once created and well tested.

If there is a native flutter fix for supporting system fonts in canvaskit, that would definitely be optimal, since for example, the [Noto Sans JP Google Font](https://fonts.google.com/specimen/Noto+Sans+JP?subset=japanese) is 4.9MB for a single text style. This download size is too much.

**Comment by creativecreatorormaybenot** [comment]
@YowFung I suggest you create a new issue for this. It seems like a valid concern but will not receive any triage as a comment in this issue.
